straxen/contexts.py

```diff
@@ -307,24 +307,80 @@ def xenonnt_led(**kwargs):
     return st
 
 
+def xenonnt_simulation_offline(output_folder: str = './strax_data',
+                               wfsim_registry: str = 'RawRecordsFromFaxNT',
+                               run_id: ty.Optional[str] = None,
+                               global_version: ty.Optional[str] = None,
+                               fax_config: ty.Optional[str] = None,
+                               ):
+    """
+    :param output_folder: strax_data folder
+    :param wfsim_registry: Raw_records generation mechanism, 'RawRecordsFromFaxNT', 'RawRecordsFromMcChain', etc,
+                           https://github.com/XENONnT/WFSim/blob/master/wfsim/strax_interface.py
+    :param run_id: Real run_id to use to fetch the corrections
+    :param global_version: Global versions
+                           https://github.com/XENONnT/corrections/tree/master/XENONnT/global_versions
+    :param fax_config: WFSim configuration files
+                       https://github.com/XENONnT/private_nt_aux_files/blob/master/sim_files/fax_config_nt_sr0_v4.json
+    :return: strax context for simulation
+    """
+
+    if run_id is None:
+        raise Exception("Specify a run_id to load the corrections")
+    if global_version is None:
+        raise Exception("Specify a correction global version")
+    if fax_config is None:
+        raise Exception("Specify a simulation configuration file")
+
+    import wfsim
+    # General strax context, register common plugins
+    st = strax.Context(torage=strax.DataDirectory(output_folder),
+                       **straxen.contexts.xnt_common_opts)
+    # Register simulation configs required by WFSim plugins
+    st.config.update(dict(detector='XENONnT',
+                          fax_config=fax_config,
+                          check_raw_record_overlaps=True,
+                          **straxen.contexts.xnt_common_config))
+    # Register WFSim raw_records plugin to overwrite real data raw_records
+    wfsim_plugin = getattr(wfsim, wfsim_registry)
+    st.register(wfsim_plugin)
+    for plugin_name in wfsim_plugin.provides:
+        assert 'wfsim' in str(st._plugin_class_registry[plugin_name])
+    # Register offline global corrections same as real data
+    st.apply_xedocs_configs(version=global_version)
+    # Real data correction is run_id dependent, but in sim we often use run_ids not in the runDB
+    # So we switch the run_id dependence to a specific run -> run_id
+    local_versions = st.config
+    for config_name, url_config in local_versions.items():
+        if isinstance(url_config, str):
+            if 'run_id' in url_config:
+                local_versions[config_name] = straxen.URLConfig.format_url_kwargs(url_config, run_id=run_id)
```

Comment:
What will happen if the `run_id` is `None`?

Reply:
@shenyangshi

Reply:
There is an if statement before to avoid this: https://github.com/XENONnT/straxen/blob/05468821e450ba5c17582cfacb860b7179d695a7/straxen/contexts.py#L328-L329